runtime/callback.c

```diff
@@ -19,26 +19,51 @@
 
 #include <string.h>
 #include "caml/callback.h"
-#include "caml/domain.h"
+#include "caml/codefrag.h"
 #include "caml/fail.h"
+#include "caml/fiber.h"
 #include "caml/memory.h"
 #include "caml/mlvalues.h"
+#include "caml/platform.h"
+
+/*
+ * These macros are for ensuring effects are handled correctly
+ * inside callbacks. There are two aspects:
+ *  - we clear the stack parent for a callback to force an Unhandled
+ *  exception rather than effects being passed over the callback
+ *  - we register the stack parent as a local root while the callback
+ * is executing to ensure that the garbage collector follows the
+ * stack parent
+ */
+Caml_inline value save_and_clear_stack_parent(caml_domain_state* domain_state) {
+  struct stack_info* parent_stack = Stack_parent(domain_state->current_stack);
+  value cont = caml_alloc_1(Cont_tag, Val_ptr(parent_stack));
+  Stack_parent(domain_state->current_stack) = NULL;
+  return cont;
+}
+
+Caml_inline void restore_stack_parent(caml_domain_state* domain_state,
+                                      value cont) {
+  struct stack_info* parent_stack = Ptr_val(Op_val(cont)[0]);
+  CAMLassert(Stack_parent(domain_state->current_stack) == NULL);
+  Stack_parent(domain_state->current_stack) = parent_stack;
+}
+
+#define RESTORE_STACK_PARENT(domain_state) \
+
 
 #ifndef NATIVE_CODE
 
 /* Bytecode callbacks */
 
-#include "caml/codefrag.h"
 #include "caml/interp.h"
 #include "caml/instruct.h"
 #include "caml/fix_code.h"
-#include "caml/stacks.h"
-
-CAMLexport int caml_callback_depth = 0;
+#include "caml/fiber.h"
 
-static opcode_t callback_code[] = { ACC, 0, APPLY, 0, POP, 1, STOP };
+static __thread opcode_t callback_code[] = { ACC, 0, APPLY, 0, POP, 1, STOP };
```

Comment:
Is it necessary that this is data structure is thread-local? Does the lazy initialisation of callback code on the first call to `caml_callbackN_exn` justify the cost of doing this once per thread?

Reply:
It is necessary to be thread-local as the code stands. In `caml_callbackN_exn` this array is mutated with:
```C
  callback_code[1] = narg + 3;
  callback_code[3] = narg;
```
multiple domains must therefore not use the same `callback_code` array. 
To pursue a stack local scheme for `callback_code`, wouldn't you need to do `caml_register_code_fragment` and `caml_thread_code` on every callback and also a `caml_remove_code_fragment` before the stack goes out of scope. 

Arguments in favour of the current code are:
- it will preserve the runtime performance profile for callbacks (no malloc/free and skiplist insert/remove for the codefrag accounting)
- it matches the existing trunk structure

Reply:
This makes sense to me. Thanks!